camlp5/Camlp5Helpers.ml

```diff
@@ -596,42 +573,29 @@ module Sig = struct
     let tdDef =
       (* TODO: error about gadts may be here *)
       let cs =
-        List.map (fun (name,t) -> (loc, VaVal name, VaVal [], Some t, <:vala< [] >>) )
+        List.map (fun (name,t) -> <:constructor< $uid:name$ : $t$ >> )
           constructors in
-      (TySum (loc, VaVal cs))
+      <:ctyp< [ $list:cs$ ] >>
     in
-    let td =
-      { tdNam = VaVal (loc, VaVal name);
-        tdPrm = VaVal (List.init params_count (fun n ->
-            (VaVal (Some (Printf.sprintf "dummy%d" n)), None)) );
-        tdPrv = VaVal false;
-        tdDef;
-        tdCon = VaVal [] ;
-        tdIsDecl = true ;
-        tdAttributes = <:vala< [] >>
-      }
+    let tdPrm = List.init params_count (fun n ->
+            (VaVal (Some (Printf.sprintf "dummy%d" n)), None)) in
+    let td = <:type_decl< $tp:(loc, VaVal name)$ $list:tdPrm$ = $tdDef$ >>
     in
     <:sig_item< type $list:[td]$ >>
 
 
 
   let tdecl_abstr: loc:loc -> string -> string option list -> t = fun ~loc name params ->
 
-    let td =
-      { tdNam = VaVal (loc, VaVal name);
-        tdPrm = VaVal (List.map (fun s -> (VaVal s,None)) params);
-        tdPrv = VaVal false;
-        tdDef = <:ctyp< 'abstract >>;
-        tdCon = VaVal [] ;
-        tdIsDecl = true ;
-        tdAttributes = <:vala< [] >>
-      }
+    let tdPrm = List.map (fun s -> (VaVal s,None)) params in
+    let td = <:type_decl< $tp:(loc, VaVal name)$ $list:tdPrm$ = 'abstract >>
     in
     <:sig_item< type $list:[td]$ >>
 
   let module_ ~loc (_,name,mtyp) =
     <:sig_item< module $uid:name$ : $mtyp:mtyp$ >>
 
+  (* TODO: Kakadu, what is this?  I don't recognize this.  It doesn't seem to exist in Ocaml, and the construct appears to be something leftover in camlp5. *)
   let modtype ~loc (_loc,s,mt_opt) =
```

Comment:
It is module type decaration in signature. I does exist in upstream OCaml
https://github.com/ocaml/ocaml/blob/trunk/parsing/parsetree.mli#L767
Can you explain what confused you?

Reply:
Ha!  Nice!  Yet another reason why I'm glad you gave me GT to hack on!  It turns out, I added a -different- syntax for abstract module-types, b/c I didn't realize this one existed!  I'm removing the other one, so this code will continue to work unmodified.  *grin*

Reply:
OK, I removed that superfluous AST syntax, so what you've got works, as it should, and as it did before.